mbed_build/_internal/config/config_layer.py

```diff
@@ -43,16 +43,36 @@ def from_config_source(cls, config_source: ConfigSource, target_labels: List[str
 
         This method translates data found in ConfigSource into modifiers specific to the given target.
         """
-        modifiers_from_config = [
-            build_modifier_from_config_entry(key=key, data=data) for key, data in config_source.config.items()
-        ]
+        namespace = config_source.namespace
+        modifiers = []
+        for key, data in config_source.config.items():
+            modifiers.append(build_modifier_from_config_entry(key=_namespace(key, namespace), data=data))
 
-        modifiers_from_target_overrides: List[Callable] = []
         allowed_target_labels = ["*"] + target_labels
         for target_label, overrides in config_source.target_overrides.items():
             if target_label in allowed_target_labels:
-                modifiers_from_target_overrides.extend(
-                    build_modifier_from_target_override_entry(key=key, data=data) for key, data in overrides.items()
-                )
+                for key, data in overrides.items():
+                    modifiers.append(
+                        build_modifier_from_target_override_entry(key=_namespace(key, namespace), data=data)
+                    )
 
-        return cls(config_source=config_source, modifiers=(modifiers_from_config + modifiers_from_target_overrides))
+        return cls(config_source=config_source, modifiers=modifiers)
+
+
+def _namespace(key: str, namespace: str) -> str:
+    """Prefix configuration key with a namespace.
+
+    Namespace is ConfigSource wide, and is resolved at source build time.
+
+    It should be one of:
+    - "target"
+    - "application"
+    - library name (where "mbed_lib.json" comes from)
+
+    If given key is already namespaced, return it as is - this is going to be the case for
+    keys from "target_overrides" entries. Keys from "config" usually need namespacing.
```

Comment:
Wow, so target overrides in the mbed_lib and mbed_app files don't just override target keys? They can also override lib and app keys but still be called target overrides?

Reply:
Yes. So far the case where this was happening I've found was `mbed_lib.json` file that defined defaults and then used `"target_overrides"` to have per-target specific overrides.
https://github.com/ARMmbed/mbed-os/blob/master/platform/mbed_lib.json

I'm pretty sure there's some validation step that prevents you from overriding configuration specified by other lib, but I'm unable to dig it out right now and as said I'd rather validate separately and just focus on parsing right now. :)

Reply:
From what I understand about this (this explanation is probably going to make little sense as my understanding of this is sketchy, please correct me if any of this is wrong), the `target_overrides` can override "app specific" or "module specific" defs (set in app.json or lib.json respectively in the `config` section), to build for a specific target. Arbitrary Example: an app sets the `stack_size` to some nominal default value, but certain targets may not support that so they override it in `target_overrides`. On top of that the `target_overrides` can also override "base" target specific defines (I assume this is the wonderful stuff in `targets.json`) to build a certain app or module. For example a target could set its stack size to `0x1` by default, but a Wi-Fi module may need a smaller/larger stack size than that so it would override `target.stack_size` in that case (I'm just pulling the example defines and modules out of thin air so the names and values may not be accurate)... So yeah this is a rather messy system. This config rubix cube makes CMake toolchain files seem like a walk in the park (and I always thought CMake handled toolchain specific build settings poorly).

Reply:
@rwalton-arm this does match my understanding as well.